### ISSUE TITLE
AcceleratedOptimizer.eval() does not reach through DeepSpeed wrapper, silently saving train-view schedule-free weights

### ISSUE THREAD

**Issue by XIONGPEILIN** [huggingface/accelerate#4159]
### System Info

```shell
- Accelerate version: 1.14.0
- Platform: Linux 5.4.0 x86_64, glibc 2.31
- Python version: 3.13.11
- NumPy version: 2.4.6
- PyTorch version: 2.12.0+cu130
- PyTorch accelerator: CUDA
- DeepSpeed version: 0.19.1
- Mixed precision: bf16
- Observed with DeepSpeed ZeRO-1 and a schedule-free optimizer
```

### Information

- [ ] The official example scripts
- [x] My own modified scripts

### Tasks

- [ ] One of the scripts in the examples folder
- [x] My own task or dataset: distributed diffusion-transformer training with DeepSpeed ZeRO-1 and ProdigyPlusScheduleFree

### Reproduction

`AcceleratedOptimizer.train()` knows how to reach through the extra optimizer wrapper introduced by DeepSpeed, but `AcceleratedOptimizer.eval()` only checks the immediate wrapped object. If that DeepSpeed wrapper does not expose `eval()`, the call silently becomes a no-op.

The wrapper asymmetry can be reproduced on CPU without installing DeepSpeed:

```python
import torch
from accelerate import Accelerator
from accelerate.optimizer import AcceleratedOptimizer


class ScheduleFreeLikeOptimizer(torch.optim.SGD):
    def __init__(self, *args, **kwargs):
        super().__init__(*args, **kwargs)
        self.mode = None

    def train(self):
        self.mode = "train"

    def eval(self):
        self.mode = "eval"


class DeepSpeedLikeOptimizerWrapper:
    """Models the extra optimizer layer used by DeepSpeed."""

    def __init__(self, optimizer):
        self.optimizer = optimizer

    def state_dict(self):
        return self.optimizer.state_dict()

    def load_state_dict(self, state_dict):
        self.optimizer.load_state_dict(state_dict)


Accelerator()
model = torch.nn.Linear(10, 10)
inner = ScheduleFreeLikeOptimizer(model.parameters(), lr=0.1)
optimizer = AcceleratedOptimizer(DeepSpeedLikeOptimizerWrapper(inner))

optimizer.train()
assert inner.mode == "train"

optimizer.eval()
assert inner.mode == "eval"  # Fails: inner.mode is still "train"
```

Current `train()` contains a DeepSpeed-specific fallback:

```python
elif (
    hasattr(self.optimizer, "optimizer")
    and hasattr(self.optimizer.optimizer, "train")
    and callable(self.optimizer.optimizer.train)
):
    self.optimizer.optimizer.train()
```

The corresponding fallback is absent from `eval()`.

This is particularly dangerous for schedule-free optimizers because it does not raise an exception. In an actual bf16 ZeRO-1 checkpoint, the saved optimizer metadata still contained:

```text
use_schedulefree=True
train_mode=True
```

after calling `optimizer.eval()` immediately before `accelerator.save_state()`. Both the sharded fp32 master and the bf16 module state therefore remained in the training view, so a later inference checkpoint conversion silently recovered the wrong parameters.

This appears to be the missing symmetric half of #3266, which added DeepSpeed wrapper traversal to `train()` only. A matching fix and CPU regression tests are already proposed in #4130.

### Expected behavior

`AcceleratedOptimizer.eval()` should reach through the same one-level DeepSpeed optimizer wrapper as `train()`:

```python
def eval(self):
    if hasattr(self.optimizer, "eval") and callable(self.optimizer.eval):
        self.optimizer.eval()
    elif (
        hasattr(self.optimizer, "optimizer")
        and hasattr(self.optimizer.optimizer, "eval")
        and callable(self.optimizer.optimizer.eval)
    ):
        self.optimizer.optimizer.eval()
```

After `optimizer.eval()`, the underlying schedule-free optimizer must report `train_mode=False`, allowing `accelerator.save_state()` to capture its evaluation-view fp32 master weights. Calling `optimizer.train()` after saving should continue to restore training mode as it does today.

**Comment by SunMarc** [comment]
fixed
